File: business/src/main/java/no/nav/sbl/dialogarena/soknadinnsending/business/db/SQLUtils.java
package no.nav.sbl.dialogarena.soknadinnsending.business.db;

import static java.lang.System.getProperty;

public class SQLUtils {
	
	public static final String DIALECT_PROPERTY = "sqldialect";

	public static String limit(int limit) {
		if ("hsqldb".equals(getProperty(DIALECT_PROPERTY))) {
			return "limit " + limit;
		} else {
			return "and rownum <= " + limit;
		}
	}

	public static String whereLimit(int limit) {
		if ("hsqldb".equals(getProperty(DIALECT_PROPERTY))) {
			return "limit " + limit;
		} else {
			return "where rownum <= " + limit;
		}
	}

	public static String toDate(int antallDager) {
		if ("hsqldb".equals(getProperty(DIALECT_PROPERTY))) {
			return "CURRENT_TIMESTAMP - " + antallDager + " DAY";
		} else {
			return "CURRENT_TIMESTAMP - NUMTODSINTERVAL("+antallDager+",'DAY') " ;
		}
	}

	public static String selectNextSequenceValue(String sequence) {
		if ("hsqldb".equals(getProperty(DIALECT_PROPERTY))) {
			return "call next value for " + sequence;
		} else {
			return "select " + sequence + ".nextval from dual";
		}
	}

	public static String selectMultipleNextSequenceValues(String sequence) {
		if ("hsqldb".equals(getProperty(DIALECT_PROPERTY))) {
			return "select next value for " + sequence + " from unnest(sequence_array(1,?,1))";
		} else {
			return "select " + sequence + ".nextval from dual connect by level <= ?";
		}
	}

}

File: src/main/kotlin/no/nav/sosialhjelp/soknad/db/SQLUtils.kt
package no.nav.sosialhjelp.soknad.db

import java.sql.Timestamp
import java.time.LocalDateTime

object SQLUtils {

    private const val HSQLDB = "hsqldb"
    private const val POSTGRESQL = "postgresql"
    const val DIALECT_PROPERTY = "sqldialect"

    fun limit(limit: Int): String {
        return when (System.getProperty(DIALECT_PROPERTY)) {
            HSQLDB, POSTGRESQL -> "limit $limit"
            else -> "and rownum <= $limit"
        }
    }

    fun selectNextSequenceValue(sequence: String): String {
        return when (System.getProperty(DIALECT_PROPERTY)) {
            HSQLDB -> "call next value for $sequence"
            POSTGRESQL -> "select nextval('$sequence')"
            else -> "select $sequence.nextval from dual"
        }
    }

    fun tidTilTimestamp(tid: LocalDateTime?): Timestamp? {
        return tid?.let { Timestamp.valueOf(it) }
    }

    fun nullableTimestampTilTid(timestamp: Timestamp?): LocalDateTime? {
        return timestamp?.toLocalDateTime()
    }

    fun timestampTilTid(timestamp: Timestamp): LocalDateTime {
        return timestamp.toLocalDateTime()
    }

    fun blobSizeQuery(): String {
        return when (System.getProperty(DIALECT_PROPERTY)) {
            HSQLDB, POSTGRESQL -> "OCTET_LENGTH(DATA)"
            else -> "dbms_lob.getLength(DATA)"
        }
    }
}
